include/alpaka/kernel/TaskKernelCpuOmp2Blocks.hpp

```diff
@@ -38,12 +39,694 @@
 #    include <stdexcept>
 #    include <tuple>
 #    include <type_traits>
+#    include <utility>
 #    if ALPAKA_DEBUG >= ALPAKA_DEBUG_MINIMAL
 #        include <iostream>
 #    endif
 
 namespace alpaka
 {
+    namespace detail
+    {
+        //! Executor of parallel OpenMP loop with the given schedule
+        //!
+        //! Is explicitly specialized for all supported schedule kinds to help code optimization by compilers.
+        //! For the same purpose performs dispatch based on ways of setting chunk size.
+        //!
+        //! \tparam TKernel The kernel type.
+        //! \tparam TSchedule The schedule type (not necessarily omp::Schedule).
+        //! \tparam TScheduleKind The schedule kind value.
+        template<typename TKernel, typename TSchedule, omp::Schedule::Kind TScheduleKind, typename TSfinae = void>
+        struct ParallelForImpl;
+
+        //! Executor of parallel OpenMP loop with no schedule set
+        //!
+        //! Does not use chunk size.
+        //!
+        //! \tparam TKernel The kernel type.
+        //! \tparam TSchedule The schedule type (not necessarily omp::Schedule).
+        template<typename TKernel, typename TSchedule>
+        struct ParallelForImpl<TKernel, TSchedule, omp::Schedule::NoSchedule>
+        {
+            //! Run parallel OpenMP loop
+            //!
+            //! \tparam TLoopBody The loop body functor type.
+            //! \tparam TIdx The index type.
+            //!
+            //! \param loopBody The loop body functor instance, takes iteration index as input.
+            //! \param numIterations The number of loop iterations.
+            template<typename TLoopBody, typename TIdx>
+            ALPAKA_FN_HOST void operator()(
+                TKernel const&,
+                TLoopBody&& loopBody,
+                TIdx const numIterations,
+                TSchedule const&)
+            {
+#    if _OPENMP < 200805 // For OpenMP < 3.0 you have to declare the loop index (a signed integer) outside of the loop
```

Comment:
I would also suggest to remove the #if's here and go with the OPENMP<200805 branch only.
In addition you could create two macros, which take the schedule kind as an argument. One macro without chunksize and one with chunksize. Then the impls get one liner (somewhat difficult to read and debug, but this is alpaka, there is a lot difficult to read and debug). That's up to the alpaka team.

Reply:
> I would also suggest to remove the #if's here and go with the OPENMP<200805 branch only.

In terms of shortening the code, which is also now copied 10 times I agree. However this may be against the spirit of optimization, which is the only reason for this PR? The other branch allows using an actual data type and not wrapping it into the definitely wide enough type.

With the macros I actually had the same idea initially, but then decided it's not worth it. But ultimately i think it does not make a big difference here: not much simplification, not much less readability.

Reply:
@psychocoderHPC what do you think about the suggestion to remove `if`-guards?

Reply:
I am not sure, keep it for now and we should discuss this in the alpaka dev meeting. Maybe can find a more elegant way.
I would like to drop `200805` but I heard that Microsoft is only providing OMP2.

Reply:
maybe we should add a macro as @krzikalla suggested

```C++
#    if _OPENMP < 200805 // For OpenMP < 3.0 you have to declare the loop index (a signed integer) outside of the loop
                         // header.
#define ALPAKA_OMP_LOOP(iterations, ...)  \
                std::intmax_t iNumBlocksInGrid(static_cast<std::intmax_t>(iterations)); \
                std::intmax_t i; \
                __pragma(omp for nowait __VA_ARGS__) \
                for(i = 0; i < iNumBlocksInGrid; ++i) 
#    else
#define ALPAKA_OMP_LOOP(iterations, ...)  \
                __pragma(omp for nowait __VA_ARGS__) \
                for(TIdx i = 0; i < iterations; ++i) \
#    endif
```

But if I would do it in a separate PR.

Reply:
Microsoft has started adding support for OpenMP see here: https://devblogs.microsoft.com/cppblog/improved-openmp-support-for-cpp-in-visual-studio/

As we only support the latest greatest Visual Studio version we could try to use the new command line switch to enable it.

Reply:
MS is in a mix of OpenMP 2.0 with some OpenMP 3.0 features. They also allow using the LLVM OpenMP runtime now. It definitely looks like there is some movement recently: https://devblogs.microsoft.com/cppblog/improved-openmp-support-for-cpp-in-visual-studio/
But for the time being, it's still OpenMP 2.0.